alphagov/govuk-knowledge-graph-search
# Fix auth & cors error when running a search and container is fresh

# Problem
Our application was intermittently encountering CORS errors during keyword search operations, specifically when the server container was freshly initiated without any active user sessions.

# Origin of the Problem
This error originated from our authentication server, Signon, which implements the OAuth2.0 protocol. When an unauthenticated AJAX search request was made, the backend would redirect the request to Signon. However, Signon's CORS policy blocked AJAX requests from our server, causing the CORS error.

 # Solution
I have made updates to distinguish between browser and AJAX requests and handle them differently:

- Introduced `x-requested-with` header in frontend AJAX requests to differentiate them from browser requests.
- Updated the backend authentication to handle browser and AJAX requests differently:
    - For AJAX requests, return a 401 status.
    - For browser requests, redirect to Signon.
- Added a new middleware to the backend for authentication and a utility function to detect AJAX requests. The new middleware replaces the `connect-ensure-login` package we were using.
- Implemented an update to the frontend to refresh the page if a fetch request returns a 401 status, leading to a browser request and thus allowing the backend to redirect the user to Signon.

# Tests POC

I've added a small TS unit tests suite using jest, to test the util. I'm aware we didn't have tests yet. So this isn't an opinionated move, and I'm happy to hear any suggestions on this.


# How to test
1. Open govsearch in dev
2. in a new tab, open gov.uk signon portal
3. Log out of signon whilst keeping the govsearch tab open
4. Restart the running container CloudRun. It can be achieved my moving 100% of the traffic to an old revision, and then moving it back to the new revision.
5. Run a search in gov search
6. You should be redirected to signon. Upon login, you should be redirected to govsearch.

Before the fix, you wouldn't be redirected to signon, but you would get a cors error in the devtools console instead.
